JDBC Step3 띠용

jye326 opened the pull request:
안녕하세요 한스! 데모데이 준비로 텀이 좀 길었습니다 ㅎㅎ

~service 코드가 지저분해져서 '이렇게 하는 게 맞나..?' 하는 생각이 자꾸 드네요.. 자유롭게 피드백 부탁드려요.~

`TransactionManager`를 도입하니 서비스 코드가 좀 깔끔해졌네요. 근데 이제 Service에서 DataSource를 어떻게 생성해야 할지가 고민입니다.

jye326 committed:
refactor: JdbcTemplate과 UserHistoryDao 변경 및 코드 개선

- JdbcTemplate에 getDataSource 메서드 추가
- UserHistoryDao 생성자에서 JdbcTemplate의 getDataSource 활용
- try-with-resources 문법으로 리소스 관리 개선
- 불필요한 finally 블록 제거 및 코드 간결화
- UserServiceTest의 @Disabled 제거
- UserService에 비밀번호 변경 트랜잭션 적용 로직 주석 추가

jye326 committed:
refactor: UserDao 및 UserService에서 트랜잭션 처리 로직 추가 및 JdbcTemplate 개선

- UserDao에 Connection을 사용하는 update 메서드 추가
- UserService에서 비밀번호 변경 시 트랜잭션 처리 로직 추가
- JdbcTemplate에 Connection을 지원하는 update 메서드 구현
- UserHistoryDao에 Connection 기반 로그 작성 메서드 추가
- 안전한 롤백 메서드(safeRollback) 구현
- 예외 처리 방식 개선 및 코드 가독성 향상

jye326 commented:
@coderabbitai 리뷰해줘~

jye326 committed:
refactor: UserHistoryDao의 JdbcTemplate 활용으로 코드 간결화

- PreparedStatement 및 Connection 관련 코드 제거
- JdbcTemplate의 update 메서드 사용으로 중복 제거
- 리소스 관리 개선 및 가독성 향상

jye326 committed:
refactor: UserService에 safeRollback 로깅 추가

- safeRollback 메서드에서 Connection null 체크 로직 추가
- Connection이 null인 경우 디버그 로그 출력

jye326 committed:
refactor: UserService에서 예외 타입 변경으로 예외 처리 통일성 개선

- RuntimeException 대신 DataAccessException으로 변경
- 예외 타입 명확화로 코드 가독성 향상

jye326 commented:
@coderabbitai 다시 리뷰

jye326 commented:
@coderabbitai 리뷰 좀 해 달라니깐

jye326 committed:
refactor: UserService 및 UserHistoryDao 리팩토링으로 가독성 및 일관성 개선

- UserDao에 getConnection 메서드 추가로 중복 제거
- UserService에서 UserDao의 getConnection 활용 및 트랜잭션 코드 단축
- UserHistoryDao의 SQL 상수화로 코드 재

jye326 committed:
feat: TransactionManager 도입으로 트랜잭션 관리 리팩토링

- TransactionManager 및 TransactionalAction 인터페이스 추가
- UserService에 TransactionManager 적용
- 트랜잭션 로직 캡슐화로 코드 중복 제거 및 안정성 강화

jye326 committed:
refactor: 불필요한 import 제거 및 코드 스타일 수정

- 사용되지 않는 Reflections 관련 import 제거
- 예외 메시지 문자열 formatter 통일
- 코드 명료성과 가독성 향상

jye326 committed:
refactor: 예외 메시지 정리로 코드 가독성 및 명확성 개선

- DataAccessException 및 RuntimeException의 예외 메시지 형식 수정
- 메시지에 Exception 객체를 포함하여 디버깅 정보 강화

20HyeonsuLee commented on app/src/main/java/com/techcourse/service/UserService.java:
```diff
@@ -1,18 +1,30 @@
     public UserService(final UserDao userDao, final UserHistoryDao userHistoryDao) {
         this.userDao = userDao;
         this.userHistoryDao = userHistoryDao;
+        DataSource dataSource = userDao.getDataSource();
```
DataSourceConfig.getInstance()로 가져오는건 어떨까요!?

jye326 replied:
좋은데요? 반영했습니다!

20HyeonsuLee commented on app/src/main/java/com/techcourse/service/UserService.java:
```diff
@@ -26,7 +38,9 @@ public void insert(final User user) {
         user.changePassword(newPassword);
-        userDao.update(user);
-        userHistoryDao.log(new UserHistory(user, createBy));
+        transactionManager.executeTransaction(connection -> {
+            userDao.update(connection, user);
+            userHistoryDao.log(connection, new UserHistory(user, createBy));
+        });
```
트랜잭션을 하나의 작업 단위로 묶어서 한번에 실행하도록 설계하셨군요👍

20HyeonsuLee commented on jdbc/src/main/java/com/interface21/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,31 @@
+                action.execute(connection);
+                connection.commit();
+            } catch (Exception e) {
+                connection.rollback();
+                throw new DataAccessException("트랜잭션 오류발생으로 롤백", e);
```
롤백 과정에서 예외가 발생하면, 기존 예외가 덮어씌워지는 문제가 발생할 수 있을 것 같아요!

20HyeonsuLee commented on jdbc/src/main/java/com/interface21/transaction/TransactionManager.java:
```diff
@@ -0,0 +1,31 @@
+            connection.setAutoCommit(false);
+            try {
+                action.execute(connection);
+                connection.commit();
```
추후에 해당 커넥션을 이용하는 곳에서는 connection.setAutoCommit(false);인 상태로 이용하게 될 것 같습니다!

commit후에 true로 바꿔주는건 어떨까요?

jye326 replied:
오 좋네요!

20HyeonsuLee reviewed (approved):
안녕하세요 띠용~ 이번 step도 잘 구현해 주셨군요!

DataSourceConfig에 DataSource가 static으로 관리되고 있어서 jdbcTemplate을 통해서 getDataSource를 하지 않아도 될 것 같아요.

Service에서는 DataSourceConfig.getInstance를 활용해서 DataSource를 얻어오면 될 것 같습니다!

(jdbcTemplate.getDataSource와 DataSourceConfig.getInstance는 같은 객체를 반환함)

레벨4가 얼마 남지 않았기 때문에 Approve하도록 하겠습니다. 리뷰 확인 후 띠용이 직접 Merge해주세요!

jye326 committed:
refactor: 트랜잭션 및 DataSource 초기화 로직 개선

- TransactionManager에 autoCommit 설정 추가로 롤백 후 안정성 강화
- UserService에서 DataSourceConfig를 활용한 DataSource 초기화로 의존성 관리 개선
